compiler/src/main/scala/org/scalajs/core/compiler/ExplicitLocalJS.scala

```diff
@@ -0,0 +1,412 @@
+            atPos(tree.pos) {
+              localTyper.typed {
+                gen.mkMethodCall(Runtime_newInnerOrLocalJSObject,
+                    List(newTpt.tpe), List(superJSClassValue, newCall))
+              }
+            }
+          }
+
+        /* Wrap `super` calls to inner and local JS classes with
+         * `innerOrLocalJSSuperCall`, to preserve a reified reference to the
+         * necessary JS class value.
+         */
+        case Apply(fun @ Select(sup: Super, _), _)
+            if !fun.symbol.isConstructor &&
+                isInnerOrLocalJSClass(sup.symbol.superClass) =>
+          val newSuperCall = super.transform(tree)
+          val superJSClassValue =
+            genJSConstructorOf(tree, sup.symbol.superClass.tpe_*)
+          atPos(tree.pos) {
+            localTyper.typed {
+              gen.mkMethodCall(Runtime_innerOrLocalJSSuperCall,
+                  List(newSuperCall.tpe), List(superJSClassValue, newSuperCall))
+            }
+          }
+
+        // Same for a super call with type parameters
+        case Apply(TypeApply(fun @ Select(sup: Super, _), _), _)
+            if !fun.symbol.isConstructor &&
+                isInnerOrLocalJSClass(sup.symbol.superClass) =>
+          val newSuperCall = super.transform(tree)
+          val superJSClassValue =
+            genJSConstructorOf(tree, sup.symbol.superClass.tpe_*)
+          atPos(tree.pos) {
+            localTyper.typed {
+              gen.mkMethodCall(Runtime_innerOrLocalJSSuperCall,
+                  List(newSuperCall.tpe), List(superJSClassValue, newSuperCall))
+            }
+          }
+
+        // Translate js.constructorOf[T]
+        case Apply(TypeApply(ctorOfTree, List(tpeArg)), Nil)
+            if ctorOfTree.symbol == JSPackage_constructorOf =>
+          val newTpeArg = transform(tpeArg)
+          gen.mkAttributedCast(genJSConstructorOf(tree, newTpeArg.tpe),
+              JSDynamicClass.tpe)
+
+        // Translate x.isInstanceOf[T] for inner and local JS classes
+        case Apply(TypeApply(fun @ Select(obj, _), List(tpeArg)), Nil)
+            if fun.symbol == Any_isInstanceOf &&
+                isInnerOrLocalJSClass(tpeArg.tpe.typeSymbol) =>
+          val newObj = transform(obj)
+          val newTpeArg = transform(tpeArg)
+          atPos(tree.pos) {
+            localTyper.typed {
+              val jsCtorOf = genJSConstructorOf(tree, newTpeArg.tpe)
+              gen.mkMethodCall(Special_instanceof, List(newObj, jsCtorOf))
+            }
+          }
+
+        case _ =>
+          super.transform(tree)
+      }
+    }
+
+    /** Generates the desugared version of `js.constructorOf[tpe]`.
+     */
+    private def genJSConstructorOf(tree: Tree, tpe: Type): Tree = {
+      val clazz = tpe.typeSymbol
+
+      if (clazz.isTrait || clazz.isModuleClass) {
+        throw new TypeError(tree.pos,
+            s"non-trait class type required but $tpe found")
```

Comment:
Because it should have been checked earlier in PrepJSInterop. If we arrive here and we fall in that case, it's a compiler bug (or a bug of another compiler plugin trying to mess with us).